### PR TITLE
New filename scheme based on external id instead of project name.

### PR DESCRIPTION
**Pull request by taikuukaits**
This makes slightly longer filenames in most cases (group, artifact, version) instead of (artifact, version) but should reduce conflicts. I've also improved the long filename ordering after talking with brian.

### CODE REVIEW

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy [hunk 1]
```diff
@@ -193,27 +197,39 @@ class DetectProjectManager implements SummaryResultReporter {
...
-        def names = [finalSourcePathPiece, projectName, projectVersionName]
-        names.sort { -it.size() }
-        String filename = generateFilename(bomToolType, finalSourcePathPiece, projectName, projectVersionName)
-        for (int i = 0; (filename.length() >= 255) && (i < 3); i++) {
-            names[i] = DigestUtils.sha1Hex(names[i])
-            if (names[i].length() > 15) {
-                names[i] = names[i].substring(0, 15)
-            }
+    private String generateShortenedFilename(BomToolType bomToolType, String finalSourcePathPiece, ExternalId externalId) {
+
```
Comment (ekerwin) [inline src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy]: gah

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy [hunk 2]
```diff
@@ -193,27 +197,39 @@ class DetectProjectManager implements SummaryResultReporter {
...
+        rawPieces.add(bomToolType.toString());
+        rawPieces.addAll(pieces);
+        rawPieces.add('bdio');
 
-        String filename = (safePieces as Iterable).join('_') + '.jsonld'
-        filename
+        List<String> safePieces = rawPieces.collect { integrationEscapeUtil.escapeForUri(it) };
+
+        String filename = (safePieces as Iterable).join('_') + '.jsonld';
+
```
Comment (ekerwin) [inline src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy]: I know Iterable was here before...do we know why? It seems odd to me - I thought join was available to a List. Am I off base here?

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy [hunk 3]
```diff
@@ -193,27 +197,39 @@ class DetectProjectManager implements SummaryResultReporter {
...
-            }
+    private String generateShortenedFilename(BomToolType bomToolType, String finalSourcePathPiece, ExternalId externalId) {
+
+        List<String> filenamePieces = new ArrayList<>();
+        filenamePieces.addAll(externalId.getExternalIdPieces());
+        filenamePieces.add(finalSourcePathPiece);
+        String filename = generateFilename(bomToolType, filenamePieces)
 
-            filename = generateFilename(bomToolType, names[0], names[1], names[2])
+        if (filename.length() >= 255){
```
Comment (ekerwin) [inline src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy]: Nice - sorting this direction makes it easier to have an arbitrary number of pieces. Solid.

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy (hunk 2 above)
Comment (taikuukaits) [reply src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy]: When I try to remove that it says:
Deprecated.  Use the Iterable version of join instead

File: src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy (hunk 1 above)
Comment (taikuukaits) [reply src/main/groovy/com/blackducksoftware/integration/hub/detect/DetectProjectManager.groovy]: Fixed this - what other whitespace were you substantially objecting to?

**Pull request merged**